source/common/unified_memory_allocation/src/memory_provider.c

```diff
@@ -28,37 +29,112 @@ umaMemoryProviderCreate(struct uma_memory_provider_ops_t *ops, void *params,
 void umaMemoryProviderDestroy(uma_memory_provider_handle_t hProvider) {
-    hProvider->ops.finalize(hProvider->provider_priv);
+    hProvider->ops.finalize(hProvider->priv);
     free(hProvider);
 }
 
 enum uma_result_t umaMemoryProviderAlloc(uma_memory_provider_handle_t hProvider,
                                          size_t size, size_t alignment,
                                          void **ptr) {
-    return hProvider->ops.alloc(hProvider->provider_priv, size, alignment, ptr);
+    return hProvider->ops.alloc(hProvider->priv, size, alignment, ptr);
 }
 
 enum uma_result_t umaMemoryProviderFree(uma_memory_provider_handle_t hProvider,
                                         void *ptr, size_t size) {
-    return hProvider->ops.free(hProvider->provider_priv, ptr, size);
+    return hProvider->ops.free(hProvider->priv, ptr, size);
 }
 
 enum uma_result_t
 umaMemoryProviderGetLastResult(uma_memory_provider_handle_t hProvider, const char **ppMessage) {
-    return hProvider->ops.get_last_result(hProvider->provider_priv, ppMessage);
+    return hProvider->ops.get_last_result(hProvider->priv, ppMessage);
+}
+
+struct uma_tracking_memory_provider_t {
+    uma_memory_provider_handle_t hUpstream;
+    uma_memory_tracker_handle_t hTracker;
+    void *pool;
+};
+
+typedef struct uma_tracking_memory_provider_t uma_tracking_memory_provider_t;
+
+static enum uma_result_t trackingAlloc(void *hProvider,
+                                       size_t size, size_t alignment,
+                                       void **ptr) {
+    uma_tracking_memory_provider_t *p = (uma_tracking_memory_provider_t *)hProvider;
+
+    enum uma_result_t ret = umaMemoryProviderAlloc(p->hUpstream, size, alignment, ptr);
+    if (ret != UMA_RESULT_SUCCESS) {
+        return ret;
+    }
+
+    ret = umaMemoryTrackerAdd(p->hTracker, p->pool, *ptr, size);
+    if (ret != UMA_RESULT_SUCCESS) {
+        // TODO: free the allocation and return error?
+    }
+
+    return ret;
+}
+
+static enum uma_result_t trackingFree(void *hProvider,
+                                      void *ptr, size_t size) {
+    uma_tracking_memory_provider_t *p = (uma_tracking_memory_provider_t *)hProvider;
+
+    enum uma_result_t ret = umaMemoryProviderFree(p->hUpstream, ptr, size);
+    if (ret != UMA_RESULT_SUCCESS) {
+        return ret;
+    }
+
+    ret = umaMemoryTrackerRemove(p->hTracker, ptr, size);
+    if (ret != UMA_RESULT_SUCCESS) {
+        // TODO
+    }
+
+    return ret;
+}
+
+static enum uma_result_t trackingInitialize(void *params, void **ret) {
+    uma_tracking_memory_provider_t *provider = (uma_tracking_memory_provider_t *)malloc(sizeof(uma_tracking_memory_provider_t));
+    *provider = *((uma_tracking_memory_provider_t *)params);
+    *ret = provider;
+    return UMA_RESULT_SUCCESS;
+}
+
+static void trackingFinalize(void *provider) {
+    free(provider);
+}
+
+static struct uma_memory_provider_ops_t umaTrackingMemoryProviderOps() {
+    struct uma_memory_provider_ops_t ops = {
```

Comment:
I don't think you need to wrap this in a function. Just create a static variable.